--- orient.py	2023-05-16 20:34:42.000000000 +0530
+++ orient_new.py	2023-07-31 15:18:28.781884000 +0530
@@ -190,12 +190,15 @@
 def read_xyz(filename):
     '''reads xyz file and returns Geometry object'''
     out = []
-
+    line_counter=0
     with open(filename, "r") as f:
-        line = f.readline()
-        while line != "":
-            natoms = int(line)
+        line = f.readline().strip()
+        natoms= int(line)
+        
+        while line != "" and line_counter != natoms:
+            #natoms = int(line)
             comment = f.readline().rstrip()
+            
 
             names = []
             coords = []
@@ -203,6 +206,7 @@
 
             for i in range(natoms):
                 line = f.readline()
+                line_counter+=1
                 data = line.split()
                 name, x, y, z = data[0:4]
                 extra = data[4:]
@@ -211,10 +215,11 @@
                 coords.append( [float(x), float(y), float(z)] )
                 if extra:
                     extras.append(extra)
-
+		 
             out.append(Geometry(names, np.array(coords), comment=comment, extras=extras))
 
             line = f.readline()
+       
 
     return out
 
